toolchain/check/handle_function.cpp

```diff
@@ -212,8 +212,11 @@ static auto BuildFunctionDecl(Context& context,
       parent_scope_inst) {
     if (auto class_decl = parent_scope_inst->TryAs<SemIR::ClassDecl>()) {
       auto& class_info = context.classes().Get(class_decl->class_id);
-      CARBON_CHECK(virtual_modifier != SemIR::Function::VirtualModifier::Impl ||
-                   class_info.is_dynamic);
+      if (virtual_modifier == SemIR::Function::VirtualModifier::Impl &&
+          !class_info.base_id.is_valid()) {
+        CARBON_DIAGNOSTIC(ImplWithoutBase, Error, "impl without base class");
+        context.emitter().Build(node_id, ImplWithoutBase).Emit();
```

Comment:
I suppose we'll eventually want to check for a *dynamic* base class here, and then check that the base class has a matching virtual function. This seems like a good step towards that.

Reply:
Yep. Given your comment I thought about whether I could/should add the check that the base was dynamic in this patch - but thinking about it, I'm not sure we'd want to describe that to a user (it'd amount to "there are no virtual functions in the base class", I think? - which is not a distinctly interesting property for the user, probably, and would fall out of/probably use the same diagnostic as "we couldn't find a virtual function with this name in any base" (that there were /no/ virtual functions at all doesn't seem like an interesting distinction))